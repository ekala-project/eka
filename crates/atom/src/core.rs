//! # Atom Core Types
//!
//! This module contains the fundamental types that represent atoms and their
//! file system structure. These types form the foundation of the atom format
//! and are used throughout the crate.

use std::collections::HashMap;
use std::path::{Path, PathBuf};

use semver::Version;
use serde::{Deserialize, Serialize};

use super::id::{Label, Name};
use crate::manifest::AtomSets;

//================================================================================================
// Types
//================================================================================================

/// Represents the deserialized form of an Atom, directly constructed from the TOML manifest.
///
/// This struct contains the basic metadata of an Atom but lacks the context-specific
/// [`crate::AtomId`], which must be constructed separately.
#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct Atom {
    /// The verified, human-readable Unicode identifier for the Atom.
    pub label: Label,

    /// The version of the Atom.
    pub version: Version,

    /// An optional description of the Atom.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    /// A table of named atom sets, defining the sources for resolving atom dependencies.
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub sets: HashMap<Name, AtomSets>,
}

/// Represents the file system paths associated with an atom.
///
/// This struct manages the relationship between an atom's manifest file
/// (the "spec") and its content directory. It handles the logic for determining
/// these paths based on whether we're given a manifest file or a content directory.
#[derive(Debug)]
pub(crate) struct AtomPaths<P>
where
    P: AsRef<Path>,
{
    /// Path to the atom's manifest file (atom.toml)
    spec: P,
    /// Path to the atom's content directory
    content: P,
}

//================================================================================================
// Impls
//================================================================================================

impl AtomPaths<PathBuf> {
    /// Creates a new `AtomPaths` instance from a given path.
    ///
    /// If the path points to a manifest file (named `atom.toml`), then:
    /// - `spec` is set to that file path
    /// - `content` is set to the parent directory
    ///
    /// If the path points to a directory, then:
    /// - `spec` is set to `path/atom.toml`
    /// - `content` is set to the provided path
    ///
    /// # Arguments
    ///
    /// * `path` - Either a path to a manifest file or content directory
    ///
    /// # Returns
    ///
    /// An `AtomPaths` instance with the appropriate spec and content paths.
    pub(crate) fn new<P: AsRef<Path>>(path: P) -> Self {
        let path = path.as_ref();
        let name = path
            .file_name()
            .unwrap_or(path.as_os_str())
            .to_string_lossy();

        if name == crate::ATOM_MANIFEST_NAME.as_str() {
            AtomPaths {
                spec: path.into(),
                content: path.parent().unwrap_or(Path::new("")).into(),
            }
        } else {
            let spec = path.join(crate::ATOM_MANIFEST_NAME.as_str());
            AtomPaths {
                spec: spec.clone(),
                content: path.into(),
            }
        }
    }

    /// Returns the path to the atom's manifest file.
    ///
    /// This is the `atom.toml` file that contains the atom's metadata
    /// and dependency specifications.
    pub fn spec(&self) -> &Path {
        self.spec.as_ref()
    }

    /// Returns the path to the atom's content directory.
    ///
    /// This directory contains the actual source code or files that
    /// make up the atom's content.
    pub fn content(&self) -> &Path {
        self.content.as_ref()
    }
}
