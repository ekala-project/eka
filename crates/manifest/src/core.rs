use atom::{Atom, Id};
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use thiserror::Error;
use toml_edit::de;
use toml_edit::DocumentMut;

#[derive(Error, Debug)]
pub enum AtomError {
    #[error("Manifest is missing the `[atom]` key")]
    Missing,
    #[error(transparent)]
    InvalidAtom(#[from] de::Error),
    #[error(transparent)]
    InvalidToml(#[from] toml_edit::TomlError),
}

type AtomResult<T> = Result<T, AtomError>;

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Manifest {
    r#trait: Id,
    pub atom: Atom,
}

impl Manifest {
    /// Get the Atom object from a toml manifest
    pub fn get_atom(content: &str) -> AtomResult<Atom> {
        let doc = content.parse::<DocumentMut>()?;

        if let Some(v) = doc.get("atom").map(|v| v.to_string()) {
            let atom = de::from_str::<Atom>(&v)?;
            Ok(atom)
        } else {
            Err(AtomError::Missing)
        }
    }
}

impl FromStr for Manifest {
    type Err = de::Error;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        de::from_str(s)
    }
}
