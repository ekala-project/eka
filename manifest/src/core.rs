use super::Name;
use semver::Version;
use serde::{Deserialize, Serialize};
use std::str::FromStr;
use toml_edit::de;
use toml_edit::DocumentMut;

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Manifest {
    r#trait: Name,
    pub atom: Atom,
}

#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub struct Atom {
    id: Name,
    version: Version,
    #[serde(skip_serializing_if = "Option::is_none")]
    description: Option<String>,
}

impl Manifest {
    pub fn is(content: &str) -> bool {
        let doc = match content.parse::<DocumentMut>() {
            Ok(doc) => doc,
            Err(_) => return false,
        };

        if let Some(v) = doc.get("atom").and_then(|v| v.as_str()) {
            if de::from_str::<Atom>(v).is_ok() {
                return true;
            }
        }

        false
    }
}

impl FromStr for Manifest {
    type Err = de::Error;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        de::from_str(s)
    }
}
