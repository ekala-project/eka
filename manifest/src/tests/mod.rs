use super::core::Manifest;

#[test]
fn serialize() -> anyhow::Result<()> {
    let results: Vec<Manifest> = vec![
        r#"
            trait = "package"
            [atom]
            id = "foo"
            version = "0.1.0"
        "#
        .parse()?,
        r#"
            trait = "deployment"
            [atom]
            id = "bar"
            version = "0.2.0"
        "#
        .parse()?,
        r#"
            trait = "config"
            [atom]
            id = "some"
            version = "1.2.0"
            description = "Some config"
        "#
        .parse()?,
    ];

    insta::assert_yaml_snapshot!(results);
    Ok(())
}
