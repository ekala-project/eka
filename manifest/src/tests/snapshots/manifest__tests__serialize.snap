---
source: manifest/src/tests/mod.rs
expression: results
---
- trait: package
  atom:
    id: foo
    version: 0.1.0
- trait: deployment
  atom:
    id: bar
    version: 0.2.0
- trait: config
  atom:
    id: some
    version: 1.2.0
    description: Some config
